Python/ceval.c

```diff
@@ -725,9 +726,27 @@ PyObject*
 _PyEval_MatchClass(PyThreadState *tstate, PyObject *subject, PyObject *type,
                    Py_ssize_t nargs, PyObject *kwargs)
 {
+    // Recurse on unions.
+    if (_PyUnion_Check(type)) {
+        // get union members
+        PyObject *members = _Py_union_args(type);
+        const Py_ssize_t n = PyTuple_GET_SIZE(members);
+
+        // iterate over union members and return first match
+        for (Py_ssize_t i = 0; i < n; i++) {
+            PyObject *member = PyTuple_GET_ITEM(members, i);
+            PyObject *attrs = _PyEval_MatchClass(tstate, subject, member, nargs, kwargs);
+            // match found
+            if (attrs != NULL) {
+                return attrs;
+            }
+        }
+        // no match found
+        return NULL;
+    }
     if (!PyType_Check(type)) {
-        const char *e = "called match pattern must be a class";
-        _PyErr_Format(tstate, PyExc_TypeError, e);
+        const char *e = "called match pattern must be a class or typing.Union of classes (got %s)";
```

Comment:
```suggestion
        const char *e = "called match pattern must be a class or a union of classes (got %s)";
```

I propose to be vague about the exact type of union here, given that `types.Union` and `typing.Union` used to be different and now they are the same 😱 

Also, many people might not recognize that `A | B` and `typing.Union[A, B]` is the same and use `typing.Union` directly.

Reply:
Actually, I think this probably should be changed to check all union members before recursion. Otherwise, something like this would be legal at runtime:

```pyhton
t = Sequence | "SomeNoneSenseType"
match ["a"]:
     case t():
         pass
```

Reply:
Actually, it seems isinstance supports this:

```python
isinstance(1, list[str])  # TypeError at runtime
isinstance(1, int | list[str])  # no TypeError at runtime
```

Reply:
Hm, actually the original PEP does not allow this, and this call fails in python3.10 but succeeds in python3.11+

Reply:
Note to self: This behavior was changed in python 3.11, see https://github.com/python/cpython/issues/91603